### ISSUE TITLE
GitLab sync issue

### ISSUE THREAD

**Issue by amrography** [repman-io/repman#589]
I can't sync gitlab projects or add new one.

![Screen Shot 2022-05-19 at 12 21 53 PM](https://user-images.githubusercontent.com/9614340/169272054-8fbb8009-ecdf-4461-88f5-ce5688374f40.png)
![Screen Shot 2022-05-19 at 12 22 02 PM](https://user-images.githubusercontent.com/9614340/169272061-f33bf9f9-134d-4ac1-b79b-92006b03acf6.png)

**Comment by giggsey** [comment]
What version of GitLab are you running?

I had an issue after the upgrade to GitLab 15, but somehow managed to sort it by trying to Add Projects twice until I got redirected to GitLab, then projects seemed to sync again.

**Comment by hugodeaguiar** [comment]
@giggsey It happens on the last version of gitlab.com. They published a new release that expires Oauth tokens. So every time we'll need to unlink Gitlab and link again to refresh the token.

**Comment by hugodeaguiar** [comment]
Take a look: https://about.gitlab.com/blog/2022/04/18/gitlab-releases-15-breaking-changes/#oauth-tokens-without-an-expiration

**Comment by kristofnk** [comment]
Can you enhance the token refresh mechanism in repman so we can access gitlab again?

**Comment by amrography** [comment]
@giggsey Yes, I am doing this now. Thanks for the tip.

**Comment by giggsey** [comment]
> @giggsey Yes, I am doing this now. Thanks for the tip.

I need to retest, as I expect it'll be erroring again due to the token expiring

**Comment by Fahl-Design** [comment]
any updates on this?

**Comment by tecbird** [comment]
+1

**Comment by Fahl-Design** [comment]
FYI: I found the root cause, the "redirect_uri" must always be the same when requesting a new token with a refresh token

- at registration redirect_uri `.../register/gitlab/check` is used
- when this token expires and an UpdateMessage is trigger a refresh a token is requested with redirect_uri: `../user/token/gitlab/check`
- this will result in:
```json
{
  "error": "invalid_grant",
  "error_description": "The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."
}
```

**Comment by KSauter** [comment]
Hi @Fahl-Design, can you write a small summary how to fix it in a running container? Would be very helpful until this issue is fixed in repman itself.

**Comment by Fahl-Design** [comment]
@KSauter yes, my boss may found an other hotfix:
- remove current token(s) from database table "user_oauth_token"
- relink the user you want to use
- open `vendor/league/oauth2-client/src/Provider/AbstractProvider.php` and goto `\League\OAuth2\Client\Provider\AbstractProvider::getAccessToken`
- force `.../register/gitlab/check` url for "refresh_token" requests like: 
```php
if ($params['grant_type'] === 'refresh_token') {
  $params['redirect_uri'] = 'https://<YOUR_REPMAN_URI>/register/gitlab/check';
}
```
to verify:
- change expire date in database and try to update a package

**Comment by KSauter** [comment]
@Fahl-Design Thanks for the update!
I updated the expires_at to be in the past, and the update was successful without any code change 

`update user_oauth_token set expires_at = '2022-06-20 10:10:10' where id = 'xxx';`

**Comment by Fahl-Design** [comment]
@KSauter yes it will work when the token was generated with redirect_uri "/user/token/gitlab/check" (which is the case when you generate the user token by adding a package. when you link the account first, it will fail later

you can see it with some hacky style debug logging (remove tmp fix first)

```php
public function getAccessToken($grant, array $options = [])
    {
        $grant = $this->verifyGrant($grant);

        $params = [
            'client_id'     => $this->clientId,
            'client_secret' => $this->clientSecret,
            'redirect_uri'  => $this->redirectUri,
        ];

        $params   = $grant->prepareRequestParameters($params, $options);
// tmp fix
if ($params['grant_type'] === 'refresh_token') {
  $params['redirect_uri'] = 'https://<YOUR_REPMAN_URI>/register/gitlab/check';
}
        try {
            file_put_contents('/tmp/repman.log', var_export($params, true) ."\n".var_export($options, true) , FILE_APPEND | LOCK_EX);
            $request  = $this->getAccessTokenRequest($params);
        } catch (\Throwable $e) {
            dd($e);
        }

        $response = $this->getParsedResponse($request);
        if (false === is_array($response)) {
            throw new UnexpectedValueException(
                'Invalid response received from Authorization Server. Expected JSON.'
            );
        }
        $prepared = $this->prepareAccessTokenResponse($response);
        $token    = $this->createAccessToken($prepared, $grant);

        return $token;
    }
```

**Comment by KSauter** [comment]
@Fahl-Design i checked further and found that the refresh-token isn't updated by the token refresh. This will invalidate the refresh token for the next use.

See PR https://github.com/repman-io/repman/pull/596
